File: Ruby/Civitas/pkg/Civitas-0.0.1/lib/mazo_sorpresas.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

# encoding: utf-8
require_relative 'diario'
require_relative 'sorpresa'

module Civitas
  class MazoSorpresas
    def initialize
      @sorpresas=Array.new
      @barajda=false
      @usadas=0
      @debug=false
      @cartas_especiales=Array.new
      @ultima_sorpresa=nil
      
      if(@debug==true)
        mensaje="Debug = true"
        Diario.instance.ocurre_evento(mensaje)
      end
    end
    
    def self.nuevo_mazo(db)
      new()
      @debug=db
    end
    
    attr_accessor :sorpresas, :barajada, :usadas, :debug, :ultima_sorpresa, :cartas_especiales

    def al_mazo(carta)
      if(barajada==false)
        sorpresas.push(carta)
      end
    end
    
    def siguiente
      if(barajada==false || usadas==sorpresas.length)
        if(!debug)
          @usadas=0
          sorpresas.shuffle
          @barajada=true
        end
      end
      
      @usadas+=1
      siguiente=sorpresas.first
      sorpresas.delete_at(0)
      sorpresas.push(siguiente)
      @ultima_sorpresa=siguiente
      
      return siguiente
    end
    
    def inhabilitar_carta_especial(sorpresa)
      for i in (0..@sorpresas.length)
        if(@sorpresas[i]==sorpresa)
          sorpresas.delete(sorpresa)
          cartas_especiales.push(sorpresa)
          mensaje="Se ha retirado la sorpresa del mazo"
          Diario.instance.ocurre_evento(mensaje)
        end
      end
    end
    
    def habilitar_carta_especial(sorpresa)
      for i in (0..@cartas_especiales.length)
        if(@cartas_especiales[i]==sorpresa)
          cartas_especiales.delete(sorpresa)
          sorpresas.push(sorpresa)
          mensaje="Se ha retirado la sorpresa del mazo"
          Diario.instance.ocurre_evento(mensaje)
        end
      end
    end
    
    def get_ultima_sorpresa
      ultima=@ultima_sorpresa
      
      return ultima
    end
  end
end

File: Java/P5/Civitas/src/civitas/mazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author arturo
 */
public class mazoSorpresas {
    private ArrayList<Sorpresa> sorpresas; 
    private boolean barajada; 
    private int usadas; 
    private boolean Debug; 
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa; 
    
    private void init(){
        barajada=false;
        usadas=0;
        sorpresas=new ArrayList<>();
        cartasEspeciales=new ArrayList<>();
    }
    
    mazoSorpresas(){
        Debug=false;
        this.init();
    }
    
    mazoSorpresas(boolean d){
          String event="Debug activado";
          Debug=d;
          this.init();
    }
    
    void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }
    
    int getTamMazo(){
        return sorpresas.size();
    }
    
    Sorpresa getSorpresas(int i){
        return sorpresas.get(i);
    }

    Sorpresa siguiente(){
        boolean aux=false;
        
        if(!barajada || usadas==sorpresas.size()){
            if(!Debug){
                usadas=0;
                Collections.shuffle(sorpresas);
                barajada=true;
                aux=true; 
            }
        }
        
        usadas++;
        ultimaSorpresa=sorpresas.get(0);
        sorpresas.add(ultimaSorpresa);
        sorpresas.remove(0);
        
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        boolean aux=false;
        
        for(int i=0; i<sorpresas.size(); i++){
            if(sorpresas.contains(sorpresa)){
                sorpresas.remove(i);
                cartasEspeciales.add(sorpresa);
                aux=true;
            }
        }
        
        aux=false;
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        boolean aux=false;
        boolean aux2=false;
        
        for(int i=0; i<cartasEspeciales.size(); i++){
            if(cartasEspeciales.contains(sorpresa)){
                cartasEspeciales.remove(i);
                sorpresas.add(sorpresa);
                Diario.getInstance().ocurreEvento("Carta especial habilitada");
            }
        }
    }
}
